### PR TITLE
tests for CASSANDRA-8838

### PR DESCRIPTION
**Pull request by yukim**
[CASSANDRA-8838](https://issues.apache.org/jira/browse/CASSANDRA-8838) introduced new feature around bootstrappping/replace node.
Tests are added to verify its functionality.

### CODE REVIEW

**Comment by ptnapoleon** [comment]
I'm +1, these are sophisticated tests, nice job. If I could make one nit, it's that the comments on `InterruptBootstrap` calls should explicitly state that the node being killed is doing outbound streaming. It would make it much clearer.

**Comment by ptnapoleon** [comment]
Looking for a second review, @aboudreault or @EnigmaCurry ?

File: bootstrap_test.py [hunk 1]
```diff
@@ -76,3 +80,80 @@ def read_from_bootstrapped_node_test(self):
...
+        t = InterruptBootstrap(node1)
+        t.start()
+
+        # start bootstrapping node3 and wait for streaming
+        node3 = new_node(cluster)
+        node3.set_configuration_options(values={'stream_throughput_outbound_megabits_per_sec': 1})
+        try:
+            node3.start()
+        except NodeError:
+            pass # node doesn't start as expected
```
Comment (aboudreault) [inline bootstrap_test.py]: Can the node3 start for some reason? If so, is it wrong? Assuming so, could we ensure the expected exception with assertRaises ?

File: bootstrap_test.py [hunk 2]
```diff
@@ -76,3 +80,80 @@ def read_from_bootstrapped_node_test(self):
...
+        t.join()
```
Comment (aboudreault) [inline bootstrap_test.py]: same comment than above...

File: bootstrap_test.py (hunk 1 above)
Comment (ptnapoleon) [inline bootstrap_test.py]: I think the issue is the node fails to start due to the failed bootstrap, right @yukim? `NodeError` is a ccm problem. The logs are still parsed at the end of the test for unexpected errors, and Yuki updated the expected log errors at the top of the test.

**Comment by aboudreault** [comment]
Just a question about a try/except in the code. Otherwiset, LGTM. (y)

**Comment by yukim** [comment]
ccm throws NodeError at the end of ```node#start()``` if the node is not running. (see https://github.com/pcmanus/ccm/blob/master/ccmlib/node.py#L508)

I am intentionally killing the node to let bootstrap fail, so the node throws NodeError from ```node#start()``` and I need to catch them to keep test running.

**Comment by ptnapoleon** [comment]
Are good to merge then, @aboudreault ?

**Comment by aboudreault** [comment]
Yes!

**Pull request merged**